torch/_functorch/aot_autograd.py

```diff
@@ -3357,6 +3655,264 @@ def forward(*runtime_args):
+                        o.detach_()
+                out_loss = out[output_loss_index]
+                num_fw_outs = len(out)
+            else:
+                out_loss = out
+                num_fw_outs = 1
+            assert isinstance(out_loss, torch.Tensor)
+            assert out_loss.requires_grad
+            assert out_loss.numel() == 1, "When passing a module into aot_export_module, the output must be a scalar loss."
+            return out
+        ctx = nullcontext
+    else:
+        # Run under no_grad, so our tracing machinery only traces an inference graph.
+        ctx = torch.no_grad
+        fn_to_trace = functional_call
+
+    full_args = []
+    # First, the params
+    full_args.extend(params_and_buffers_flat)
+    # Next, the input args
+    full_args.extend(args)
+
+    with ctx():
+        fx_g, metadata, in_spec, out_spec = _aot_export_function(
+            fn_to_trace,
+            full_args,
+            decompositions=decompositions,
+            num_params_buffers=len(params_and_buffers_flat),
+        )
+    if trace_joint:
+        def flattened_joint(*args):
+            # The idea here is that the joint graph that AOTAutograd creates has some strict properties:
+            # (1) It accepts two arguments (primals, tangents), and pytree_flattens them
+            # (2) It returns a tuple of (fw_outs, gradients)
+            # This is a very useful convention for anyone who wants to partition the joint graph
+            # into a separate forward and backward graph.
+            # However,
+            # (1) for people exporting a single joint graph, it would be preferable not to have
+            #     any pytrees in the graph.
+            # (2) We are guaranteed in the aot_export_module case that the forward outputs a loss,
+            #     and there are therefore no tangents that are needed to run the joint graph.
+            # (3) AOTAutograd creates a grad_input for every input in the forward,
+            #     including None's for inputs that are not grad-requiring tensors.
+            #     we don't want these in our export graph.
+            #     and there are therefore no tangents that are needed to run the joint graph.
+            # This function "fixes" both of the above by removing any tangent inputs,
+            # and removing pytrees from the original FX graph.
+            fake_tangents = [None for _ in range(metadata.num_outputs + metadata.num_mutated_inputs)]
+            fw_outs, gradients = fx_g(args, fake_tangents)
+            assert len(gradients) == len(args)
+            output_gradients = []
+            for i, (a, grad) in enumerate(zip(args, gradients)):
+                if isinstance(a, torch.Tensor) and a.requires_grad:
+                    assert grad is not None, """\
+Found a parameter that did not receive a gradient.
+"This is most likely a bug, but if this needs to be supported please file an issue on GitHub."""
```

Comment:
I like to preemptively file bugs and link 'em in the error message, so that I know I will get notified if someone does need it.

Reply:
Good call - updated, issue here https://github.com/pytorch/pytorch/issues/101192